### DISCUSSION TITLE
`ResourceReference` resource

### DISCUSSION THREAD

**Discussion by simo498** [godotengine/godot-proposals#9545] [Editor]
I'd like to discuss about a problem i've been facing using the Godot Engine.
Thers is no option to reference project files in the Engine. There is no built-in resource to reference a project file without loading it on RAM.
It can lead to various problems while developing a game (mainly on PackedScenes, which are usually very big in size):

- The resources are loaded as the node is instantiated, giving the developer less control on how and when they should be loaded (example: i could need to load them later, maybe in another thread)
- The resources will be kept in RAM for all the lifespan of the node, leading to memory usage problems
- It could lead to a dependency cycle where resources can't be loaded at all

A fix to this problem could be loading them at runtime specifying the path in the code (hardcoded), or pass the path to an exported property, but unlike resources it's impossible to keep track of moved or renamed files; you need to manually update any of the referenced paths, and this could be difficult to do (expecially when moving a large amount of files into a new folder) and can lead to null references problems.

I suggest to create a new type of resource which only keeps a String referencing a file/directory (and DOESN'T load the resource/scene automatically at node instancing); this should be able to track the file position/name and update the references when a file gets renamed or moved on the Editor (as it happens with Resources)

**Comment by Calinou** [comment]
> (mainly on PackedScenes, which are usually very big in size):

I believe there's InstancePlaceholder for this particular scenario (right-click instanced scene in the Scene tree dock > **Load as Placeholder**).

**Reply by simo498** [reply]
It's not an ideal solution in my opinion:
- I was thinking to something more general which can be used both with resources and scenes
- it instances the scene directly, not giving you the option to load it first
- it's not thread safe
